// @chalbert/enzyme-adapter-utils/build/tagCodes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TracingMarkerComponent = exports.SuspenseListComponent = exports.SuspenseComponent = exports.SimpleMemoComponent = exports.ScopeComponent = exports.Profiler = exports.OffscreenComponent = exports.Mode = exports.MemoComponent = exports.LegacyHiddenComponent = exports.LazyComponent = exports.IndeterminateComponent = exports.IncompleteClassComponent = exports.HostText = exports.HostRoot = exports.HostPortal = exports.HostComponent = exports.FunctionComponent = exports.Fragment = exports.ForwardRef = exports.DehydratedFragment = exports.ContextProvider = exports.ContextConsumer = exports.ClassComponent = exports.CacheComponent = void 0;
var FunctionComponent = 0;
exports.FunctionComponent = FunctionComponent;
var ClassComponent = 1;
exports.ClassComponent = ClassComponent;
var IndeterminateComponent = 2; // Before we know whether it is function or class

exports.IndeterminateComponent = IndeterminateComponent;
var HostRoot = 3; // Root of a host tree. Could be nested inside another node.

exports.HostRoot = HostRoot;
var HostPortal = 4; // A subtree. Could be an entry point to a different renderer.

exports.HostPortal = HostPortal;
var HostComponent = 5;
exports.HostComponent = HostComponent;
var HostText = 6;
exports.HostText = HostText;
var Fragment = 7;
exports.Fragment = Fragment;
var Mode = 8;
exports.Mode = Mode;
var ContextConsumer = 9;
exports.ContextConsumer = ContextConsumer;
var ContextProvider = 10;
exports.ContextProvider = ContextProvider;
var ForwardRef = 11;
exports.ForwardRef = ForwardRef;
var Profiler = 12;
exports.Profiler = Profiler;
var SuspenseComponent = 13;
exports.SuspenseComponent = SuspenseComponent;
var MemoComponent = 14;
exports.MemoComponent = MemoComponent;
var SimpleMemoComponent = 15;
exports.SimpleMemoComponent = SimpleMemoComponent;
var LazyComponent = 16;
exports.LazyComponent = LazyComponent;
var IncompleteClassComponent = 17;
exports.IncompleteClassComponent = IncompleteClassComponent;
var DehydratedFragment = 18;
exports.DehydratedFragment = DehydratedFragment;
var SuspenseListComponent = 19;
exports.SuspenseListComponent = SuspenseListComponent;
var ScopeComponent = 21;
exports.ScopeComponent = ScopeComponent;
var OffscreenComponent = 22;
exports.OffscreenComponent = OffscreenComponent;
var LegacyHiddenComponent = 23;
exports.LegacyHiddenComponent = LegacyHiddenComponent;
var CacheComponent = 24;
exports.CacheComponent = CacheComponent;
var TracingMarkerComponent = 25;
exports.TracingMarkerComponent = TracingMarkerComponent;
